iggy/src/users/update_user.rs

```diff
@@ -0,0 +1,234 @@
+use crate::models::user_status::UserStatus;
+use crate::users::{MAX_USERNAME_LENGTH, MIN_USERNAME_LENGTH};
+use crate::utils::text;
+use crate::validatable::Validatable;
+use bytes::BufMut;
+use serde::{Deserialize, Serialize};
+use std::fmt::Display;
+use std::str::{from_utf8, FromStr};
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Default)]
+pub struct UpdateUser {
+    #[serde(skip)]
+    pub user_id: Identifier,
+    pub username: Option<String>,
+    pub status: Option<UserStatus>,
+}
+
+impl CommandPayload for UpdateUser {}
+
+impl Validatable<Error> for UpdateUser {
+    fn validate(&self) -> Result<(), Error> {
+        if self.username.is_none() {
+            return Ok(());
+        }
+
+        let username = self.username.as_ref().unwrap();
+        if username.is_empty()
+            || username.len() > MAX_USERNAME_LENGTH
+            || username.len() < MIN_USERNAME_LENGTH
+        {
+            return Err(Error::InvalidUsername);
+        }
+
+        if !text::is_resource_name_valid(username) {
+            return Err(Error::InvalidUsername);
+        }
+
+        Ok(())
+    }
+}
+
+impl FromStr for UpdateUser {
+    type Err = Error;
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        let parts = input.split('|').collect::<Vec<&str>>();
+        if parts.is_empty() || parts.len() > 3 {
+            return Err(Error::InvalidCommand);
+        }
+
+        let user_id = parts[0].parse::<Identifier>()?;
+        let username = match parts.get(1) {
+            Some(username) => match *username {
+                "" => None,
+                _ => Some(username.to_string()),
+            },
+            None => None,
+        };
+        let status = match parts.get(2) {
+            Some(status) => match *status {
+                "" => None,
+                _ => {
+                    let status = UserStatus::from_str(status)?;
+                    Some(status)
+                }
+            },
+            None => None,
+        };
+        let command = UpdateUser {
+            user_id,
+            username,
+            status,
+        };
+        command.validate()?;
+        Ok(command)
+    }
+}
+
+impl BytesSerializable for UpdateUser {
+    fn as_bytes(&self) -> Vec<u8> {
+        let user_id_bytes = self.user_id.as_bytes();
+        let mut bytes = Vec::new();
+        bytes.extend(user_id_bytes);
+        if let Some(username) = &self.username {
+            bytes.put_u8(1);
+            bytes.put_u32_le(username.len() as u32);
+            bytes.extend(username.as_bytes());
+        } else {
+            bytes.put_u8(0);
+        }
+        if let Some(status) = &self.status {
+            bytes.put_u8(1);
+            bytes.put_u8(status.as_code());
+        } else {
+            bytes.put_u8(0);
+        }
+
+        bytes
+    }
+
+    fn from_bytes(bytes: &[u8]) -> Result<UpdateUser, Error> {
+        if bytes.len() < 5 {
+            return Err(Error::InvalidCommand);
```

Comment:
I don't like that there is no context with this erorr, but I suppose it should be fixed later on.